src/macros/macrorecorder.cpp

```diff
@@ -0,0 +1,106 @@
+#include "macrorecorder.h"
+
+#include <QtConcurrentRun>
+
+#include "control/controlproxy.h"
+#include "preferences/configobject.h"
+
+namespace {
+constexpr uint kMaxMacroSize = 1000;
+const QString kConfigGroup = QStringLiteral("[MacroRecording]");
+}
+
+MacroRecorder::MacroRecorder()
+        : m_COToggleRecording(ConfigKey(kConfigGroup, "recording_toggle")),
+          m_CORecStatus(ConfigKey(kConfigGroup, "recording_status")),
+          m_activeChannel(nullptr),
+          m_pStartRecordingTimer(this),
+          m_pRecordedActions(kMaxMacroSize) {
+    qCDebug(macroLoggingCategory) << "MacroRecorder construct";
+
+    connect(&m_COToggleRecording,
+            &ControlPushButton::valueChanged,
+            this,
+            &MacroRecorder::slotToggleRecording);
+    connect(&m_pStartRecordingTimer,
+            &QTimer::timeout,
+            this,
+            &MacroRecorder::pollRecordingStart);
+}
+
+void MacroRecorder::notifyCueJump(
+        ChannelHandle* channel, double sourceFramePos, double destFramePos) {
+    qCDebug(macroLoggingCategory) << "Jump in channel" << channel->handle();
+    if (isRecordingActive() && checkOrClaimRecording(channel)) {
+        m_pRecordedActions.emplace(sourceFramePos, destFramePos);
+        qCDebug(macroLoggingCategory) << "Recorded jump in channel" << channel->handle();
+    }
+}
+
+bool MacroRecorder::checkOrClaimRecording(ChannelHandle* channel) {
+    ChannelHandle* value = nullptr;
+    return m_activeChannel.compare_exchange_strong(value, channel) ||
+            value->handle() == channel->handle();
+}
+
+void MacroRecorder::pollRecordingStart() {
+    qCDebug(macroLoggingCategory) << "Polling for recording start";
+    if (getActiveChannel() == nullptr) {
+        return;
+    }
+    m_pStartRecordingTimer.stop();
+    m_CORecStatus.set(Status::Recording);
+}
+
+void MacroRecorder::startRecording() {
+    qCDebug(macroLoggingCategory) << "MacroRecorder recording armed";
+    m_CORecStatus.set(Status::Armed);
+    m_pStartRecordingTimer.start(300);
+}
+
+void MacroRecorder::stopRecording() {
+    qCDebug(macroLoggingCategory) << "MacroRecorder recording stop";
+    m_pStartRecordingTimer.stop();
+    m_CORecStatus.set(Status::Disabled);
+
+    ChannelHandle* channel = m_activeChannel.exchange(nullptr);
+    if (channel == nullptr) {
+        return;
+    }
+
+    emit saveMacroFromChannel(fetchRecordedActions(), *channel);
+}
+
+size_t MacroRecorder::getRecordingSize() const {
+    return m_pRecordedActions.size();
+}
+
+const ChannelHandle* MacroRecorder::getActiveChannel() const {
+    return m_activeChannel.load();
+}
+
+MacroRecorder::Status MacroRecorder::getStatus() const {
+    return Status(m_CORecStatus.get());
+}
+
+bool MacroRecorder::isRecordingActive() const {
+    return getStatus() > 0;
+}
+
+QVector<MacroAction> MacroRecorder::fetchRecordedActions() {
```

Comment:
Macro actions could be fetched periodically instead of accumulating them all until recording has been stopped.

Reply:
What would be the advantage? We could allocate a smaller queue and allow an unlimited number of actions in a queue, but on the other hand the code would become more complex and we'd need an additional thread to do that. I'm unsure if that's worth it, at least for the initial implementation.

Reply:
To ensure that the fixed-capacity queue doesn't get full. No extra thread needed, just a timer on the event loop that consumes all buffered messages when triggered and moves them into a member variable. The final invocation happens after recording and the timer have been stopped.

Not strictly needed, but then the recorded steps would be already available/visible in the UI while recording which could provide helpful feedback to the user instead of performing the recording blindfolded.

Reply:
@uklotzde do you have ideas for how to make them visible in the UI while recording?

Reply:
Oh, and I see one disadvantage:
When looping is enabled, recorded actions need to be re-enqueued immediately after stopping recording - so if they are stored separately, they also need to be fetched appropriately (from the real-time thread!).

Reply:
I think we should discuss this on Zulip, I think this might be difficult to get right. As a first idea, we could add add hotcue-like markers for the src/dest position of the last jump and additionally grey out the area between the last hotcue jumps source/dest position if src < dest. Doing this for all recorded jumps would be confusing.